pkg/rpc/response/result/invoke.go

```diff
@@ -70,22 +76,44 @@ type invokeAux struct {
 	Notifications  []state.NotificationEvent `json:"notifications"`
 	Transaction    []byte                    `json:"tx,omitempty"`
 	Diagnostics    *InvokeDiag               `json:"diagnostics,omitempty"`
+	Session        string                    `json:"session,omitempty"`
 }
 
+// iteratorInterfaceName is a string used to mark Iterator inside the InteropInterface.
+const iteratorInterfaceName = "IIterator"
+
 type iteratorAux struct {
 	Type      string            `json:"type"`
-	Value     []json.RawMessage `json:"iterator"`
-	Truncated bool              `json:"truncated"`
+	Interface string            `json:"interface,omitempty"`
+	ID        string            `json:"id,omitempty"`
+	Value     []json.RawMessage `json:"iterator,omitempty"`
+	Truncated bool              `json:"truncated,omitempty"`
 }
 
-// Iterator represents deserialized VM iterator values with a truncated flag.
+// Iterator represents VM iterator identifier. It either has ID set (for those JSON-RPC servers
+// that support sessions) or non-nil Values and Truncated set (for those JSON-RPC servers that
+// doesn't support sessions but perform in-place iterator traversing) or doesn't have ID, Values
+// and Truncated set at all (for those JSON-RPC servers that doesn't support iterator sessions
+// and doesn't perform in-place iterator traversing).
 type Iterator struct {
+	// ID represents iterator ID. It is non-nil iff JSON-RPC server support session mechanism.
+	ID *uuid.UUID
+
+	// Values contains deserialized VM iterator values with a truncated flag. It is non-nil
+	// iff JSON-RPC server does not support sessions mechanism and able to traverse iterator.
 	Values    []stackitem.Item
-	Truncated bool
+	Truncated *bool
+}
+
+// ServerIterator represents Iterator on the server side. It is not for Client usage.
```

Comment:
So why do we have it here?

Reply:
Because we unpack iterator during `Invoke` marshalling. As a consequence, iterator registration is performed as a callback inside `(r Invoke) MarshalJSON()`, so it should be able to pass iterator parameters to the server from the `result` package.